djpm/aspect_impl.go

```diff
@@ -110,213 +90,116 @@ func (aspect Aspect) GetSenderAndCallData(ctx context.Context, block int64, tx *
+	if contract == nil {
+		// not able to verify contract creation tx
+		return &types.AspectExecutionResult{
+			Gas:    gas,
+			Err:    errors.New("not able to verify contract creation tx"),
+			Revert: types.RevertTx,
 		}
 	}
-	// get binding contract address
-	contractAddr := common.HexToAddress(req.Tx.To)
-	if req.CurrInnerTx != nil && req.CurrInnerTx.To != "" {
-		contractAddr = common.HexToAddress(req.CurrInnerTx.To)
-	}
 
-	aspectCodes, err := aspect.provider.GetAccountVerifiers(req.GetTx().BlockNumber, contractAddr)
-
-	if err != nil {
-		result := types.DefJoinPointResult("transactionAdvice GetTxBondAspects error." + err.Error())
-		result.GasInfo = req.GasInfo
-		return result
-	}
-	if len(aspectCodes) == 0 {
-		result := types.DefJoinPointResult("not bond aspects.")
-		result.GasInfo = req.GasInfo
-		return result
+	aspectCodes, err := aspect.provider.GetAccountVerifiers(block, *contract)
+	if err != nil || len(aspectCodes) == 0 {
+		return &types.AspectExecutionResult{
+			Gas:    gas,
+			Err:    errors.New("contract has not bound with any verifier aspect"),
+			Revert: types.RevertTx,
+		}
 	}
 
 	// run aspects on received transaction
-	runAspect := aspect.runAspect(ctx, method, req.GasInfo.Gas, req.GetTx().BlockNumber, &contractAddr, req, aspectCodes)
-	if len(req.Tx.Hash) != 0 {
-		runAspect.TxHash = common.Bytes2Hex(req.Tx.Hash)
-	}
-	return runAspect
+	return aspect.runAspect(ctx, types.VERIFY_TX, gas, block, contract, req, aspectCodes)
 }
 
-func (aspect Aspect) runAspect(ctx context.Context, method types.PointCut, gas uint64, blockNumber int64, contractAddr *common.Address, reqData proto.Message, req []*types.AspectCode) (response *types.JoinPointResult) {
-	aspectId := ""
-	defer func() {
-		if err := recover(); err != nil {
-			// TODO log.Error(running aspect failed")
-			response.WithErr(aspectId, errors.New("fatal: panic in running aspect"))
-		}
-	}()
-
-	response = &types.JoinPointResult{}
+func (aspect Aspect) runAspect(ctx context.Context, method types.PointCut, gas uint64, blockNumber int64, contractAddr *common.Address, reqData proto.Message, aspects []*types.AspectCode) (result *types.AspectExecutionResult) {
+	result = &types.AspectExecutionResult{
+		Gas:    gas,
+		Revert: types.NotRevert,
+	}
 
-	gasLeft := gas
-	for _, aspect := range req {
-		aspectId = aspect.AspectId
-		runner, err := run.NewRunner(ctx, aspectId, aspect.Code)
+	for _, aspect := range aspects {
+		var err error
+		runner, err := run.NewRunner(ctx, aspect.AspectId, aspect.Version, aspect.Code)
 		if err != nil {
-			return response.WithErr(aspectId, err)
+			panic(err)
 		}
 
-		if res, callErr := runner.JoinPoint(method, gasLeft, blockNumber, contractAddr, reqData); callErr != nil {
-			response.WithErr(aspectId, callErr)
-		} else {
-			response.WithResponse(aspectId, res)
-		}
+		defer runner.Return()
```

Comment:
It's preferable to return the runner after use, rather than returning upon function exit. Otherwise, in scenarios with numerous bound aspects, it may lead to excessive instance allocation and potential failure of the wasm instance.

Reply:
Fixed